Added pushbullet notifications support

tperalta82 opened the pull request:
Added pushbullet notifications support

This can be improved, but right now I have it working onmy environment.
It is using the user's api keys.

ktsaou commented on conf.d/health_alarm_notify.conf:
```diff
@@ -113,6 +113,24 @@ DEFAULT_RECIPIENT_PUSHOVER=""
+# enable/disable sending pushover notifications
+SEND_PUSHBULLET="YES"
+
+# Login to pushover.net to get your pushover app token.
```
it says pushover.net is a several places, instead of pushbullet...
Is app token needed for pushbullet?

ktsaou commented on conf.d/health_alarm_notify.conf:
```diff
@@ -165,6 +183,8 @@ role_recipients_email[sysadmin]="${DEFAULT_RECIPIENT_EMAIL}"
 
 role_recipients_pushover[sysadmin]="${DEFAULT_RECIPIENT_PUSHOVER}"
 
+role_recipients_pushover[sysadmin]="${DEFAULT_RECIPIENT_PUSHBULLET}"
```
pushover again...

ktsaou commented on conf.d/health_alarm_notify.conf:
```diff
@@ -177,6 +197,8 @@ role_recipients_email[domainadmin]="${DEFAULT_RECIPIENT_EMAIL}"
 
 role_recipients_pushover[domainadmin]="${DEFAULT_RECIPIENT_PUSHOVER}"
 
+role_recipients_pushover[domainadmin]="${DEFAULT_RECIPIENT_PUSHBULLET}"
```
here too...

ktsaou commented on conf.d/health_alarm_notify.conf:
```diff
@@ -190,6 +212,8 @@ role_recipients_email[dba]="${DEFAULT_RECIPIENT_EMAIL}"
 
 role_recipients_pushover[dba]="${DEFAULT_RECIPIENT_PUSHOVER}"
 
+role_recipients_pushover[dba]="${DEFAULT_RECIPIENT_PUSHBULLET}"
```
and here...

ktsaou commented on conf.d/health_alarm_notify.conf:
```diff
@@ -203,6 +227,8 @@ role_recipients_email[webmaster]="${DEFAULT_RECIPIENT_EMAIL}"
 
 role_recipients_pushover[webmaster]="${DEFAULT_RECIPIENT_PUSHOVER}"
 
+role_recipients_pushover[webmaster]="${DEFAULT_RECIPIENT_PUSHBULLET}"
```
and here...

ktsaou commented on conf.d/health_alarm_notify.conf:
```diff
@@ -216,7 +242,8 @@ role_recipients_email[proxyadmin]="${DEFAULT_RECIPIENT_EMAIL}"
 
 role_recipients_pushover[proxyadmin]="${DEFAULT_RECIPIENT_PUSHOVER}"
 
+role_recipients_pushover[proxyadmin]="${DEFAULT_RECIPIENT_PUSHBULLET}"
```
and here...

ktsaou commented on plugins.d/alarm-notify.sh:
```diff
@@ -191,6 +197,14 @@ do
         [ "${r}" != "disabled" ] && filter_recipient_by_criticality pushover "${r}" && arr_pushover[${r/|*/}]="1"
     done
 
+    # pushover
```
and here...

ktsaou commented on plugins.d/alarm-notify.sh:
```diff
@@ -430,6 +452,40 @@ send_pushover() {
+        for user in ${userapikeys}
+        do
+            httpcode=$(${curl} --write-out %{http_code} --silent --output /dev/null \
+                 -u ""$user"": \
```
something is wrong here.
`$user` will end up uncommented.

ktsaou commented on plugins.d/alarm-notify.sh:
```diff
@@ -648,6 +704,21 @@ send_pushover "${PUSHOVER_APP_TOKEN}" "${to_pushover}" "${when}" "${goto_url}" "
+To view NetData: ${goto_url}
+The source of this alarm is line ${src}"
+pushbullet_title="${status} at ${host} ${status_message} - ${name//_/ } - ${chart}}"
+send_pushbullet "${DEFAULT_RECIPIENT_PUSHBULLET}" "$pushbullet_message" "$pushbullet_title"
```
It should not always send notifications to `${DEFAULT_RECIPIENT_PUSHBULLET}`.
It should send them to `${to_pushbullet}`

tperalta82 commented:
Wow, this is strange, i had fixed it... probably didn't commit/push, will
check it as soon as i get home

thanks a lot

On Oct 4, 2016 20:20, "Costa Tsaousis" <notifications@github.com> wrote:

> *@ktsaou* commented on this pull request.
>
> Thanks!
>
> There are a few issues. I have commented on them.
> Also, you should add something at the very end of alarm-notify.sh.
>
> To merge it, you will have to also edit the wiki and add information about
> pushbullet to it, along with a screenshot. Please be careful when editing
> the wiki.
> ------------------------------
>
> In conf.d/health_alarm_notify.conf
> <https://github.com/firehol/netdata/pull/1070#pullrequestreview-2782380>:
>
> > @@ -113,6 +113,24 @@ DEFAULT_RECIPIENT_PUSHOVER=""
>
>
>  ###############################################################################
> +
> +###############################################################################
> +# pushbullet (pushbullet.com) push notification options
> +
> +# multiple recipients can be given like this:
> +#                  "USERTOKEN1 USERTOKEN2 ..."
> +
> +# enable/disable sending pushover notifications
> +SEND_PUSHBULLET="YES"
> +
> +# Login to pushover.net to get your pushover app token.
>
> it says pushover.net is a several places, instead of pushbullet...
> Is app token needed for pushbullet?
> ------------------------------
>
> In conf.d/health_alarm_notify.conf
> <https://github.com/firehol/netdata/pull/1070#pullrequestreview-2782380>:
>
> > @@ -165,6 +183,8 @@ role_recipients_email[sysadmin]="${DEFAULT_RECIPIENT_EMAIL}"
>
>  role_recipients_pushover[sysadmin]="${DEFAULT_RECIPIENT_PUSHOVER}"
>
> +role_recipients_pushover[sysadmin]="${DEFAULT_RECIPIENT_PUSHBULLET}"
>
> pushover again...
> ------------------------------
>
> In conf.d/health_alarm_notify.conf
> <https://github.com/firehol/netdata/pull/1070#pullrequestreview-2782380>:
>
> > @@ -177,6 +197,8 @@ role_recipients_email[domainadmin]="${DEFAULT_RECIPIENT_EMAIL}"
>
>  role_recipients_pushover[domainadmin]="${DEFAULT_RECIPIENT_PUSHOVER}"
>
> +role_recipients_pushover[domainadmin]="${DEFAULT_RECIPIENT_PUSHBULLET}"
>
> here too...
> ------------------------------
>
> In conf.d/health_alarm_notify.conf
> <https://github.com/firehol/netdata/pull/1070#pullrequestreview-2782380>:
>
> > @@ -190,6 +212,8 @@ role_recipients_email[dba]="${DEFAULT_RECIPIENT_EMAIL}"
>
>  role_recipients_pushover[dba]="${DEFAULT_RECIPIENT_PUSHOVER}"
>
> +role_recipients_pushover[dba]="${DEFAULT_RECIPIENT_PUSHBULLET}"
>
> and here...
> ------------------------------
>
> In conf.d/health_alarm_notify.conf
> <https://github.com/firehol/netdata/pull/1070#pullrequestreview-2782380>:
>
> > @@ -203,6 +227,8 @@ role_recipients_email[webmaster]="${DEFAULT_RECIPIENT_EMAIL}"
>
>  role_recipients_pushover[webmaster]="${DEFAULT_RECIPIENT_PUSHOVER}"
>
> +role_recipients_pushover[webmaster]="${DEFAULT_RECIPIENT_PUSHBULLET}"
>
> and here...
> ------------------------------
>
> In conf.d/health_alarm_notify.conf
> <https://github.com/firehol/netdata/pull/1070#pullrequestreview-2782380>:
>
> > @@ -216,7 +242,8 @@ role_recipients_email[proxyadmin]="${DEFAULT_RECIPIENT_EMAIL}"
>
>  role_recipients_pushover[proxyadmin]="${DEFAULT_RECIPIENT_PUSHOVER}"
>
> +role_recipients_pushover[proxyadmin]="${DEFAULT_RECIPIENT_PUSHBULLET}"
>
> and here...
> ------------------------------
>
> In plugins.d/alarm-notify.sh
> <https://github.com/firehol/netdata/pull/1070#pullrequestreview-2782380>:
>
> > @@ -191,6 +197,14 @@ do
>          [ "${r}" != "disabled" ] && filter_recipient_by_criticality pushover "${r}" && arr_pushover[${r/|*/}]="1"
>      done
>
> +    # pushover
>
> and here...
> ------------------------------
>
> In plugins.d/alarm-notify.sh
> <https://github.com/firehol/netdata/pull/1070#pullrequestreview-2782380>:
>
> > @@ -430,6 +452,40 @@ send_pushover() {
>      return 1
>  }
>
> +# -----------------------------------------------------------------------------
> +# pushbullet sender
> +
> +send_pushbullet() {
> +    local userapikeys="${1}" message="${2}" title="${3}"  httpcode sent=0 user priority
> +
> +    if [ "${SEND_PUSHBULLET}" = "YES" -a ! -z "${userapikeys}" -a ! -z "${message}" -a ! -z "${title}" ]
> +        then
> +        #https://docs.pushbullet.com/#create-push        priority=-2
> +        for user in ${userapikeys}
> +        do
> +            httpcode=$(${curl} --write-out %{http_code} --silent --output /dev/null \
> +                 -u ""$user"": \
>
> something is wrong here.
> $user will end up uncommented.
> ------------------------------
>
> In plugins.d/alarm-notify.sh
> <https://github.com/firehol/netdata/pull/1070#pullrequestreview-2782380>:
>
> > @@ -648,6 +704,21 @@ send_pushover "${PUSHOVER_APP_TOKEN}" "${to_pushover}" "${when}" "${goto_url}" "
>  SENT_PUSHOVER=$?
>
>  # -----------------------------------------------------------------------------
> +# send the pushbullet notification
> +
> +pushbullet_message="
> +${alarm}
> +Severity: ${severity}
> +Chart: ${chart}
> +Family: ${family}
> +To view NetData: ${goto_url}
> +The source of this alarm is line ${src}"
> +pushbullet_title="${status} at ${host} ${status_message} - ${name//_/ } - ${chart}}"
> +send_pushbullet "${DEFAULT_RECIPIENT_PUSHBULLET}" "$pushbullet_message" "$pushbullet_title"
>
> It should not always send notifications to ${DEFAULT_RECIPIENT_PUSHBULLET}
> .
> It should send them to ${to_pushbullet}
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/firehol/netdata/pull/1070#pullrequestreview-2782380>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEGffuY2jfhUUECzAjhEc636Noue3vvZks5qwqb-gaJpZM4KOCTe>
> .
>

tperalta82 commented:
Screenshots on both chrome and Android

![ndata-droid](https://cloud.githubusercontent.com/assets/4300670/19109635/278a1dde-8aee-11e6-9984-0bc87a13312d.png)
![netdata-chrome](https://cloud.githubusercontent.com/assets/4300670/19109636/278b1c0c-8aee-11e6-8a09-7fc94fdbfec8.png)

ktsaou commented on plugins.d/alarm-notify.sh:
```diff
@@ -430,6 +454,42 @@ send_pushover() {
+        for user in ${recipients}
+        do
+            httpcode=$(${curl} --write-out %{http_code} --silent --output /dev/null \
+                 -u "$userapikey": \
```
I see at the pushbullet documentation, that the API Key should be given as `--header 'Access-Token: <your_access_token_here>'`.

This is a complete example from the documentation:

```sh
curl --header 'Access-Token: <your_access_token_here>' \
     --header 'Content-Type: application/json' \
     --data-binary '{"body":"Space Elevator, Mars Hyperloop, Space Model S (Model Space?)","title":"Space Travel Ideas","type":"note"}' \
     --request POST \
     https://api.pushbullet.com/v2/pushes
```

I understand that `email`, `body` and `title` can be given as POST parameters, but I don't see anything about an access token passed as the `user` parameter.

So, I trust it works the way you did it, This is just a warning that there might be something wrong, or unsupported in future releases of the pushbullet API.

ktsaou commented:
Can you point the issue at the wiki? What is wrong exactly?

ktsaou commented:
@tperalta82 I am prepared to merge it.
Review my comments, decide how is best to do it and let me know when you are ready, to merge it.

Thanks!

tperalta82 commented:
Hello @kianxineki 
Thank you for your comments, I will remove the ${status} part
As for the curl command, this was for an earlier version of the API that I had already implemented onto other monitoring solutions.
You can check it here:
https://docs.pushbullet.com/v6/
And since they keep backwards compatibility and I already had (and have it running on zabbix / other custom monitors in production servers) I didn't see a problem, but I will change this anyway, just to be updated.

As for the wiki, nevermand, github add an issue yesterday, today morning I actually was able to edit the wiki and fix the typo 
You can check my edit here:
https://github.com/firehol/netdata/wiki/health-monitoring#pushbulletcom-push-notifications

kianxineki commented:
@Tperalta82 me? I think you mean @ktsaou ;)

ktsaou commented:
@tperalta82 when you are ready for merge, let me know.

tperalta82 commented:
Hello @ktsaou 

Updated the code to be compliant with the latest pushbullet API
Fixed a typo in the config file
Added an alarm test script
You just need to run alarm-test.sh that is in the same folder as alarm-notify

It is ready for the merge, but test at your will

Thanks

ktsaou commented:
thanks! I am merging it.
